### PR TITLE
Converts all A && A.B into A?.B

### PR DESCRIPTION
**Pull request by ZeWaka**
## About The Pull Request

Implements the `?.` operator, replacing code like `A && A.B` with `A?.B`

[BYOND Ref](http://www.byond.com/docs/ref/info.html#/operator/%3f%2e):
When reading `A?.B`, it's equivalent to `A && A.B` except that `A` is only evaluated once, even if it's a complex expression like a proc call.

Semi-Mirror of https://github.com/goonstation/goonstation/pull/2413

## Why It's Good For The Game

Faster performance, test case on [gist](https://gist.github.com/ZeWaka/c65e395e779f7fc99033a2bdc9cc573a) or [Coderbus](https://discordapp.com/channels/484170914754330625/487268744419344384/765040054598369300).

Average of 9.8% - 11.1% speedup on this operation.

Future improvement: This doesn't get every case, only matches pairs, not n-combinations.

### CODE REVIEW

File: code/_onclick/item_attack.dm [hunk 1]
```diff
@@ -95,7 +95,7 @@
 	M.lastattacker = user.real_name
 	M.lastattackerckey = user.ckey
 
-	if(force && M == user && user.client)
+	if(force && M == user?.client)
```
Comment (Rohesie) [inline code/_onclick/item_attack.dm:98]: This changes behavior.

File: code/controllers/subsystem/explosions.dm [hunk 2]
```diff
@@ -250,7 +250,7 @@ SUBSYSTEM_DEF(explosions)
 			var/mob/M = MN
 			// Double check for client
 			var/turf/M_turf = get_turf(M)
-			if(M_turf && M_turf.z == z0)
+			if(M_turf?.z == z0)
```
Comment (Rohesie) [inline code/controllers/subsystem/explosions.dm:253]: This changes behavior if both `M_turf` and `z0` are `null`

File: code/controllers/subsystem/shuttle.dm [hunk 3]
```diff
@@ -404,7 +404,7 @@ SUBSYSTEM_DEF(shuttle)
 		return 1
 	var/obj/docking_port/stationary/dockedAt = M.get_docked()
 	var/destination = dockHome
-	if(dockedAt && dockedAt.id == dockHome)
+	if(dockedAt?.id == dockHome)
```
Comment (Rohesie) [inline code/controllers/subsystem/shuttle.dm:407]: May change behavior if both are `null`.

**Review by Rohesie — CHANGES_REQUESTED** [review]
Better review your mass-replacement.

**Comment by SpaceManiac** [comment]
Formally skeptical of this as they are not actually directly equivalent.

> When reading `A?.B`, it's equivalent to `A && A.B` except that `A` is only evaluated once, even if it's a complex expression like a proc call.

This is a simplification. It's actually `!isnull(A) && A.B`, which is different if A is `0` or `""`.

**Review by ZeWaka — COMMENTED** [review]
All 338 files reviewed.

**Comment by ZeWaka** [comment]
> Formally skeptical of this as they are not actually directly equivalent.
> 
> > When reading `A?.B`, it's equivalent to `A && A.B` except that `A` is only evaluated once, even if it's a complex expression like a proc call.
> 
> This is a simplification. It's actually `!isnull(A) && A.B`, which is different if A is `0` or `""`.

Wouldn't you not be able to access the member variables anyways in the case of A being `0` or `""`?

File: code/modules/surgery/surgery.dm [hunk 4]
```diff
@@ -108,7 +108,7 @@
 		var/obj/item/tool = user.get_active_held_item()
 		if(S.try_op(user, target, user.zone_selected, tool, src, try_to_fail))
 			return TRUE
-		if(tool && tool.item_flags & SURGICAL_TOOL) //Just because you used the wrong tool it doesn't mean you meant to whack the patient with it
+		if(tool?.item_flags & SURGICAL_TOOL) //Just because you used the wrong tool it doesn't mean you meant to whack the patient with it
```
Comment (AnturK) [inline code/modules/surgery/surgery.dm:111]: What SM said, take a look at this line history for example.

Comment (ZeWaka) [reply code/modules/surgery/surgery.dm:111]: I see, I'll do another pass for bitflags.

**Comment by Cyprex** [comment]
You might want to post the regex used for this as well, if applicable

File: code/modules/hydroponics/plant_genes.dm [hunk 5]
```diff
@@ -135,7 +135,7 @@
 	name = "UNKNOWN"
 
 	var/datum/reagent/R = GLOB.chemical_reagents_list[reag_id]
-	if(R && R.type == reagent_id)
+	if(R?.type == reagent_id)
```
Comment (Rohesie) [inline code/modules/hydroponics/plant_genes.dm:138]:
```suggestion
	if(R && R.type == reagent_id)
```
Old behavior should be preserved here.

File: code/modules/mob/living/carbon/human/human.dm [hunk 6]
```diff
@@ -188,7 +188,7 @@
 		dat += "</td></tr>"
 
 	var/obj/item/bodypart/O = get_bodypart(BODY_ZONE_CHEST)
-	if((w_uniform == null && !(dna && dna.species.nojumpsuit) && !(O && O.status == BODYPART_ROBOTIC)) || (obscured & ITEM_SLOT_ICLOTHING))
+	if((w_uniform == null && !(dna?.species.nojumpsuit) && !(O?.status == BODYPART_ROBOTIC)) || (obscured & ITEM_SLOT_ICLOTHING))
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/human/human.dm:191]:
```suggestion
	if((w_uniform == null && !(dna?.species.nojumpsuit) && (!O || O.status != BODYPART_ROBOTIC)) || (obscured & ITEM_SLOT_ICLOTHING))
```
This check has further problems, but it's out of scope of this PR. However, the change may affect behavior.

File: code/modules/tgui/external.dm [hunk 7]
```diff
@@ -158,7 +158,7 @@
 	// Name the verb, and hide it from the user panel.
 	set name = "uiclose"
 	set hidden = TRUE
-	var/mob/user = src && src.mob
+	var/mob/user = src?.mob
```
Comment (Rohesie) [inline code/modules/tgui/external.dm:161]: Out of scope of this PR, but this check is very weird. Then again, clients, they are arcane aplenty.

**Comment by Rohesie** [comment]
Reviewed the whole thing. The uncommented things should preserve current behavior, unless I missed something. The ones that have an issue are commented on.

If you want to go one level deeper on this little micro-optimization crusade, `length(thing)` is faster than `thing.len`, and naturally even more than `thing?.len`. The difference with the latter is that it returns `0` instead of `null` if `thing` is `null`, and it won't runtime if `thing` is of the wrong type, instead just return a falsey value.

File: code/modules/tgui/external.dm (hunk 7 above)
Comment (ZeWaka) [reply code/modules/tgui/external.dm:161]: Do you want me to revert the change? This is indeed a weird check.

**Comment by ZeWaka** [comment]
> If you want to go one level deeper on this little micro-optimization crusade, `length(thing)` is faster than `thing.len`, and naturally even more than `thing?.len`. The difference with the latter is that it returns `0` instead of `null` if `thing` is `null`, and it won't runtime if `thing` is of the wrong type, instead just return a falsey value.

We're looking at this in the counterpart PR over at goon already, yeah.

File: code/modules/tgui/external.dm (hunk 7 above)
Comment (Rohesie) [reply code/modules/tgui/external.dm:161]: Eh, keep it or revert it, up to you. Feel free to resolve this conversation.

File: code/modules/mob/living/carbon/human/human.dm (hunk 6 above)
Comment (Rohesie) [reply code/modules/mob/living/carbon/human/human.dm:191]: The `O && O.status == BODYPART_ROBOTIC` change part should be either reverted or the suggestion accepted.
We shouldn't rely on the fact that the value of `BODYPART_ROBOTIC` is currently not `null`

**Comment by LemonInTheDark** [comment]
Exactly how much faster is this change? Is it profileable?

**Comment by optimumtact** [comment]
it's a microopt so unquantifiable really, but I'm happy to merge it in the spirit of greater codebase cooperation prosperity sphere

**Comment by ZeWaka** [comment]
> Exactly how much faster is this change? Is it profileable? 

I mean it's profileable on the scale of millions of the same operation and such, as notated in the main post. I don't think you could really notice it on a SS13 level. However, it's free performance.

**Comment by LemonInTheDark** [comment]
Works for me :)

**Comment by Rohesie** [comment]
Do that one requested change and this is ready.

**Review by Rohesie — APPROVED** [review]

**Pull request merged**